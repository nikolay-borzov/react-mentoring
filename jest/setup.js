import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
