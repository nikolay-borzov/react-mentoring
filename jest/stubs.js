export const film = {
  id: 1,
  title: 'Film title',
  poster_path: 'http://via.placeholder.com/500x750',
  overview: 'Film overview',
  release_date: '2018-05-04',
  genres: ['some genre', 'another genre']
}

export const films = [film]
