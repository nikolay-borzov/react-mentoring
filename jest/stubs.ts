import { Film } from '../src/entities/film'

export const film: Film = {
  id: 1,
  title: 'Film title',
  tagline: 'Put that cookie down',
  vote_average: 7.5,
  vote_count: 667,
  release_date: '2018-05-04',
  poster_path: 'https://picsum.photos/g/300/450/?random',
  overview: 'Once upon a time...',
  budget: 15000000,
  revenue: 59735548,
  genres: ['some genre', 'another genre']
}

export const films = [film]

function getRandomImageId() {
  // https://picsum.photos/images
  return Math.floor(Math.random() * Math.floor(1084))
}

export function getRandomFilm(id: number) {
  return {
    ...film,
    id,
    poster_path: film.poster_path.replace(
      'random',
      `image=${getRandomImageId()}`
    )
  }
}
