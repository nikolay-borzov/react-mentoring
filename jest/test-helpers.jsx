import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectlyShallow = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

/**
 * Creates test closure to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {Object} params Optional params
 * @param {number} [params.testName=`contains ${componentDisplayName} component`]
 * @param {number} [params.expectedCount=1] Expected number of children components
 * @param {Object} params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  { testName, expectedCount = 1, expectedProps } = {}
) => {
  it(testName || `contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = url => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga, state = {}, arg1 = undefined) => {
  const dispatched = []
  return runSagaOriginal(
    {
      dispatch: action => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
