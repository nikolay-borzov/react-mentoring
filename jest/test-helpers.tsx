import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
