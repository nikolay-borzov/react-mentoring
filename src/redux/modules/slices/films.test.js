import { films } from '../../../../jest/stubs'
import { runSaga } from '../../../../jest/test-helpers'

import createSlice from './films'

import filmService from '../../../services/film-service'

jest.mock('../../../services/film-service')

const initialState = {
  films: {
    items: films,
    total: 3000,
    limit: 15,
    error: new Error('Generic error'),
    isFetching: true
  },
  searchParams: {
    search: '',
    searchBy: '',
    sortBy: '',
    sortOrder: '',
    limit: 1,
    filter: []
  }
}

const slice = createSlice({
  id: 'FILMS',
  rootSelector: state => state.films,
  searchParamsSelector: state => state.searchParams,
  initialState: initialState.films
})

describe('Films state slice', () => {
  describe('selectors', () => {
    const state = initialState
    const selectors = slice.selectors

    it(`returns 'items'`, () => {
      expect(selectors.films(state)).toBe(state.films.items)
    })

    it(`returns 'total'`, () => {
      expect(selectors.total(state)).toBe(state.films.total)
    })

    it(`returns 'limit'`, () => {
      expect(selectors.limit(state)).toBe(state.films.limit)
    })

    it(`returns 'isFetching'`, () => {
      expect(selectors.isFetching(state)).toBe(state.films.isFetching)
    })

    it(`returns 'error'`, () => {
      expect(selectors.error(state)).toBe(state.films.error)
    })
  })

  describe('sagas', () => {
    afterEach(() => {
      filmService.getFilms.mockClear()
    })

    it(`fetches films`, () => {
      const response = {
        data: [{ id: 1 }, { id: 2 }],
        total: 3000,
        limit: 15
      }

      const state = {
        films: {},
        searchParams: initialState.searchParams
      }

      filmService.getFilms.mockReturnValue(Promise.resolve(response))

      expect.assertions(2)
      runSaga(slice.sagas.fetchFilmsAsync, state).then(dispatched => {
        expect(dispatched).toEqual([
          slice.actionCreators.fetchFilmsRequest(),
          slice.actionCreators.fetchFilmsSuccess(response)
        ])
        expect(filmService.getFilms).toHaveBeenCalledWith(
          initialState.searchParams
        )
      })
    })

    it(`handles films fetching failure`, () => {
      const expectedError = new Error('Fetch films failure')
      filmService.getFilms.mockReturnValue(Promise.reject(expectedError))

      expect.assertions(1)
      runSaga(slice.sagas.fetchFilmsAsync).then(dispatched => {
        expect(dispatched).toEqual([
          slice.actionCreators.fetchFilmsRequest(),
          slice.actionCreators.fetchFilmsFail(expectedError)
        ])
      })
    })
  })

  describe('reducer', () => {
    const reducer = slice.reducer
    const actionTypes = slice.actionTypes

    it('returns initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot()
    })

    it(`handles 'FETCH_FILMS_REQUEST'`, () => {
      const state = { isFetching: false }
      const action = {
        type: actionTypes.FETCH_FILMS_REQUEST
      }

      expect(reducer(state, action)).toMatchObject({
        isFetching: true
      })
    })

    it(`handles 'FETCH_FILMS_SUCCESS'`, () => {
      const state = {
        isFetching: true,
        items: [],
        total: 0,
        limit: 15,
        error: 'some error'
      }
      const expectedState = {
        isFetching: false,
        items: [{ id: 1 }],
        total: 3000,
        limit: 15,
        error: null
      }

      const action = {
        type: actionTypes.FETCH_FILMS_SUCCESS,
        payload: {
          data: expectedState.items,
          total: expectedState.total,
          limit: expectedState.limit
        }
      }

      expect(reducer(state, action)).toMatchObject(expectedState)
    })

    it(`it handles 'FETCH_FILMS_FAIL'`, () => {
      const state = {
        isFetching: true,
        error: null
      }
      const expectedState = {
        isFetching: false,
        error: 'some error'
      }
      const action = {
        type: actionTypes.FETCH_FILMS_FAIL,
        payload: 'some error'
      }

      expect(reducer(state, action)).toMatchObject(expectedState)
    })
  })
})
