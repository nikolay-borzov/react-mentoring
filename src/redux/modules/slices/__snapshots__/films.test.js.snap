// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Films state slice reducer returns initial state 1`] = `
Object {
  "error": [Error: Generic error],
  "isFetching": true,
  "items": Array [
    Object {
      "genres": Array [
        "some genre",
        "another genre",
      ],
      "id": 1,
      "overview": "Film overview",
      "poster_path": "http://via.placeholder.com/500x750",
      "release_date": "2018-05-04",
      "title": "Film title",
    },
  ],
  "limit": 15,
  "total": 3000,
}
`;
